### ISSUE TITLE
WYSIWYG: Link shortcuts do not work for cross-host links

### ISSUE THREAD

**Issue by levosgien** [dotCMS/core#8229]
Ref: https://my.dotcms.com/ticket/dotcms-106

We used to use the contentAsset link for file or image links but changed that as of https://github.com/dotCMS/core/issues/7729
- Before: /contentAsset/raw-data/a7e17f67-4e09-4e51-8275-cddd9f7094b5/fileAsset
- After: /global/images/test.jpg

The problem is that all links created to files/images from other hosts are broken, where they worked before.

**Comment by wezell** [comment]
This is true for both links and images that are added in the wysiwyg

**Comment by jtesser** [comment]
I believe we addressed this in 3.3. @john-thomas-dotcms can you see if we did.  I think you were in on this

**Comment by john-thomas-dotcms** [comment]
I may be wrong here, but I tried this out and things worked for me.  Which leads me to a question (below).  First, a couple details.

The links in the latest build of 3.3 (from both the Image Editor and the info button next to a binary field) show in the form of:

 * /contentAsset/raw-data/0a6d6b08-1b29-458a-9f58-590ce8877741/image

But that's only when you link images from a binary field in the content.  Images inserted directly into the WYSIWYG from a file asset have the following format:

* /bear-mountain/images/photo-gallery/hotel9.jpg

It appears this is what you're seeing.  But as long as the file assets are available on all hosts, the images should work on all hosts.  In fact we have several documents on dotcms.com that reference file asset images this way without problems.

So my question is:

* Have the file assets being referenced in the URL been pushed to the other host?
  * If they haven't, then please push them and try the page again.
  * If they have and the URL still isn't working, then that's another issue (though I'm not seeing that in my tests - or I'm misunderstanding the problem).

**Comment by levosgien** [comment]
John-T,

Actually it's not a push pub problem, it's a cross-host linking problem.

If a file is located here:
* www.host1.com/images/1.jpg

Then a user from another host (www.test.com) adds a piece of conent and links to the image on the host1 host, the wysiwyg generates the following relative URL:
* /images/1jpg

When rendered in the front-end, this results in the image link looking like this:
* www.test.com/images/1.jpg which DOES NOT EXIST = 404 error

**Comment by joseorsini** [comment]
Steps to Reproduce. 

1. Create a blank Site.
2. Go to this Site and create a folder called /images/. Then, add an image to this folder (for testing references, it's called "IMG_5494.jpg".
3. Go back to demo.dotcms.com site and create a new Content (Generic) content. 
4. From the TinyMCE Editor, click on "Insert/Edit Image" and search the recently uploaded image.
5. The image will be added with its relative URI (/images/IMG_5494.jpg) but the image will be displayed as broken (404 error on Browser's JS Console).

<img width="1134" alt="screen shot 2016-06-03 at 16 30 09" src="https://cloud.githubusercontent.com/assets/1717110/15792018/ba03e00c-29a8-11e6-87d0-af655a586e2c.png">

**Comment by joseorsini** [comment]
Reproduced in 3.5 and 3.5.1, following the steps from my latest comment. 

Related to https://my.dotcms.com/tickets/detail.dot?id=02021d3b-afd0-4f03-8a06-615dd877d4c1

**Comment by wezell** [comment]
This commit allows you to configure what the urls inserted via the wysiwyg look like.  It does this by adding a new dotmarketing-config property `WYSIWYG_IMAGE_URL_PATTERN`.  This pattern will perform variable interpolation when inserting the url.  The default pattern is:

`WYSIWYG_IMAGE_URL_PATTERN={path}{name}?language_id={languageId}`

| Variable | replaced with |
| ------------- | ------------- |
| {name} | file name|
| {path} | folder path |
| {languageId} | the file's language |
| {hostname} | hostname for the file |
| {extension} |  file extension |
| {inode} | file inode version |
| {identifier} | file identifier |
| {shortyInode} | file 10 digit inode |
| {shortyIdentifier} | file 10 digit identifier |


## Other example and valid patterns:

### Include hostname:
`//{hostName}{path}{name}?language_id={languageId}`

### perma-linked:
`/dA/{identifier}/{name}?language_id={languageId}`

### link to specific version:
`/dA/{inode}/{name}`

### Shorty
`/dA/{shortyId}/{name}?language_id={languageId}`

**Comment by jgambarios** [comment]
PR: https://github.com/dotCMS/core/pull/12075

**Comment by bryanboza** [comment]
Fixed, tested on master // chrome // postgres

**Comment by bryanboza** [comment]
Tested on master // postgres

**Comment by john-thomas-dotcms** [comment]
Doc: https://auth.dotcms.com/docs/latest/the-wysiwyg-field#LinkFormat

**Comment by wezell** [comment]
On the doc, I would mention that the property affects image urls specifically and not href links.  It is a bit vague in the doc.  e.g.

"When you insert a link into the WYSIWYG editor... "  should read:

"When you insert n image into the WYSIWYG editor..."

**Comment by john-thomas-dotcms** [comment]
Thanks Will, I updated the doc to make it clearer
